cudaaligner/src/hirschberg_myers_gpu.cu

```diff
@@ -505,80 +483,159 @@ __device__ void hirschberg_myers_single_char_warp(int8_t*& path, int32_t* path_l
+        : data_(data)
     {
-        assert(buffer_begin_ < buffer_end_);
+        assert(buffer_begin < buffer_end);
+        if (threadIdx.x == 0 && threadIdx.y == 0 && threadIdx.z == 0)
+        {
+            data_->mutex_        = 0;
+            data_->active_warps_ = 0;
+            data_->buffer_begin_ = buffer_begin;
+            data_->cur_end_      = buffer_begin;
+            data_->buffer_end_   = buffer_end;
+        }
+        __syncthreads();
+        if (threadIdx.x % warp_size == 0)
+        {
+            atomicAdd(&data_->active_warps_, 1);
+        }
+        __syncthreads();
     }
 
-    __device__ bool inline push(T const& t, unsigned warp_mask = 0xffff'ffff)
+    __device__ bool inline push(T const& t)
     {
-        if (buffer_end_ - cur_end_ >= sizeof(T))
+        lock_mutex_high_priority();
+        bool success = false;
+        __syncwarp();
+
+        if (data_->buffer_end_ - data_->cur_end_ >= 1)
         {
-            __syncwarp(warp_mask);
-            if (threadIdx.x == 0)
+            __syncwarp();
+            if (threadIdx.x % warp_size == 0)
             {
-                std::memcpy(cur_end_, &t, sizeof(T));
+                *(data_->cur_end_) = t;
+                ++(data_->cur_end_);
             }
-            __syncwarp(warp_mask);
-            cur_end_ += sizeof(T);
-            return true;
+            success = true;
         }
         else
         {
-            if (threadIdx.x == 0)
+            if (threadIdx.x % warp_size == 0)
             {
                 printf("ERROR: stack full!");
             }
-            return false;
         }
+        __syncwarp();
+        release_mutex();
+        return success;
     }
 
-    __device__ inline void pop()
+    __device__ inline bool pop(T& element)
     {
-        assert(cur_end_ > buffer_begin_);
-        if (cur_end_ - sizeof(T) >= buffer_begin_)
-            cur_end_ -= sizeof(T);
+        lock_mutex();
+        if (threadIdx.x % warp_size == 0)
+        {
+            assert(data_->active_warps_ > 0);
+            --(data_->active_warps_);
+        }
+        release_mutex();
+        bool result = true;
+        while (1)
+        {
+            lock_mutex();
+            __syncwarp();
+            if (data_->buffer_begin_ < data_->cur_end_)
+            {
+                element = *(data_->cur_end_ - 1);
+                __syncwarp();
+                if (threadIdx.x % warp_size == 0)
+                {
+                    --(data_->cur_end_);
+                    ++(data_->active_warps_);
+                }
+                break;
+            }
+            else if (data_->active_warps_ <= 0)
+            {
+                assert(data_->buffer_begin_ == data_->cur_end_);
+                result = false;
+                break;
+            }
+            __syncwarp();
+            release_mutex();
+#if __CUDA_ARCH__ >= 700
+            asm("nanosleep.u32 100000;");
```

Comment:
why do we need a sleep command here?

Reply:
i think i get it. is it to ensure the while loop doesn't just take over an SM while spinning?

Reply:
Essentially yes, that's what this should do. It can't really take over an SM, but it wastes resources locking and unlocking the mutex all the time. In general, if there was no work just now, the probability that there's new work now is rather low... So, give it a tiny bit of time. :)
Also, while the warp sleeps, it is one warp less to fight for the atomic operations on the mutex, which will make it easier for the `high_priority_lock()` to get the lock, which in return will make it easier for a warp trying to add new work to the queue.
In general the sleep times were just a bit of trial-and-benchmark. I haven't really optimized these values (yet).